cfme/services/requests.py

```diff
@@ -1,288 +1,496 @@
 from utils.log import logger
 from utils.appliance import Navigatable
 from utils.appliance.implementations.ui import navigator, CFMENavigateStep, navigate_to
+from utils.varmeth import variable
+from utils.wait import wait_for
 
-REQUEST_FINISHED_STATES = {'Migrated', 'Finished'}
-
-buttons = Region(
-    locators=dict(
-        approve="//*[@title='Approve this Request']",
-        deny="//*[@title='Deny this Request']",
-        copy="//*[@title='Copy original Request']",
-        edit="//*[@title='Edit the original Request']",
-        delete="//*[@title='Delete this Request']",
-        submit="//span[@id='buttons_on']/a[@title='Submit']",
-        cancel="//a[@title='Cancel']",
-    )
-)
-
-request_table = Table('//*[@id="list_grid"]/table')
-fields = Region(
-    locators=dict(
-        reason=Input("reason"),
-        request_list=request_table
-    )
-)
 
-match_page = partial(match_location, controller='miq_request', title='Requests')
+class Request(Navigatable):
+    """
+    Class describes request row from Services - Requests page
+    """
 
+    REQUEST_FINISHED_STATES = {'Migrated', 'Finished'}
 
-# TODO Refactor these module methods and their callers for a proper request class
-class Request(Navigatable):
-    def __init__(self, appliance=None):
+    def __init__(self, description=None, cells=None, partial_check=False, appliance=None):
+        """
+        Args:
+            description: by default we'll be checking Description column to find required row
+            cells: cells used to find required row in table
+            partial_check: greedy search or not?
+        """
         Navigatable.__init__(self, appliance=appliance)
+        self.description = description
+        self.partial_check = partial_check
+        self.cells = {'Description': self.description} if cells is None else cells
+        self.rest = self._get_request_from_rest(self.partial_check)
+        self.row = None
+
+    @variable(alias='rest')
+    def wait_for_request(self):
+        def _finished():
+            self.rest.reload()
+            if self.rest.request_state.title() not in self.REQUEST_FINISHED_STATES:
+                return False
+            return True
+
+        wait_for(_finished, num_sec=800, delay=20, message="Request finished")
+
+    @wait_for_request.variant('ui')
+    def wait_for_request_ui(self):
+        def _finished():
+            self.update(method='ui')
+            if self.row.request_state.text not in self.REQUEST_FINISHED_STATES:
+                return False
+            return True
+
+        wait_for(_finished, num_sec=800, delay=20, message="Request finished")
+
+    def _get_request_from_rest(self, partial_check):
+        matching_requests = (self.appliance.rest_api.collections.requests.find_by(
+            description=self.cells['Description']) if partial_check is False else
+            self.appliance.rest_api.collections.requests.find_by(
+            description='%{}%'.format(self.cells['Description'])))
```

Comment:
Smth like this might be more readable:
```
searched_description = '%{}%'.format(self.cells['Description']) if partial_check else self.cells['Description']
matching_requests = self.appliance.rest_api.collections.requests.find_by(description=searched_description)
```